### PR TITLE
back off restarts of crashlooping containers

### PR DESCRIPTION
**Pull request by samsabed**
back off restarts of crashlooping containers
partial: #4121

### CODE REVIEW

**Comment by mikedanese** [comment]
@k8s-bot ok to test

**Comment by mikedanese** [comment]
@dchen1107 @yujuhong

**Comment by erictune** [comment]
document new behavior.
Maybe in docs/user-guide/pod-states.md under `RestartPolicy`.

**Comment by samsabed** [comment]
ok will update pod-states.md 
I should have also included a description with the PR ! 
The backoff is 0, 1*SyncFrequency, 2*SyncFrequency, 4*, 8*... capped at 5mins resets after 10mins
It's applicable to containers that have terminated only, not restarted ones eg liveness failure.
Changes to the container spec resets the backoff to 0
Not sure how to report the events just to say "backing off..." or include the amount currently it's 
BackOff restarting failed docker container for : 10s
BackOff restarting failed docker container for : 20s

It's a small change on top of this to backoff pulling images.

**Comment by erictune** [comment]
Backoffs seems reasonable.  

There should be an e2e test about backoffs, but it will be so slow that it needs to run in a special queue.  @ixdy @quinton-hoole what is the right way to write a e2e test that intentionally takes tens of minutes to run?

**Comment by erictune** [comment]
Not sure how much value events are in this case, since we already have events for starting it, and the user can see that the starting interval is changing.  Also, if we do have an event, I don't like having the backoff quantity in the event because it hinders coalescing of similar events.

I think maybe the right thing to do here is to add a "backed-off" condition to pods, which kubelet sets when it is being backedoff.  We could extend kubectl describe to highlight pods that are backed-off.
@bgrant0607 are you okay with such a condition?

**Comment by dchen1107** [comment]
This is one of intents I initially introduced api.ContainerStateWaiting. Today only reason of putting a container to Waiting state is pulling image, backing-off could be another reason.

**Comment by quinton-hoole** [comment]
@erictune We don't yet have a special queue for intentionally slow e2e tests, but this seems like a good time to create one.   We do currently have a parallel e2e test run (where a subset of e2e tests are run in parallel).  Currently that takes 22 min to build and run (~8 min to build, ~12 min to run), so if your slow e2e test takes less than 12 min to run, it will not in practice slow things down if added to that queue.

Tests that should in general be skipped are defined here:

https://github.com/kubernetes/kubernetes/blob/36f94fabbe945e2fd72551c8783552fef7ab230a/hack/jenkins/e2e.sh#L88

So add your new test there.

The tests currently excluded from that parallel run are defined here:

https://github.com/kubernetes/kubernetes/blob/36f94fabbe945e2fd72551c8783552fef7ab230a/hack/jenkins/e2e.sh#L103

... so just make sure that your new test is not included there and it will run in parallel.

**Comment by quinton-hoole** [comment]
Er, actually no, I lied.  Skipped tests are skipped, so the above won't quite work.   Get back to me once you have the test ready, and we can figure out something.

**Comment by dchen1107** [comment]
@samsabed I briefly chatted with @bgrant0607. Both of us think for now we can reset Container to api.ContainerStateWaiting state with proper reason for backoff. Please note that 'reason' should be short and unique, and should be CamelCase. Thanks!

**Comment by ixdy** [comment]
The easiest way (right now) to make sure a test is run basically nowhere is to put `[Skipped]` in the name of the spec. Then you can create a specific Jenkins project which focuses only that test (or others like it).

Perhaps using a tag named something like `[Slow]` would be more meaningful in this case; it could also be added to the default skip list in e2e.sh.

**Comment by samsabed** [comment]
Thanks all for the very helpful input
I've updated the PR for ContainerStateWaiting, event compression friendly events and pod-states.md will follow up on the image pull back-off in another PR. I haven't completed the e2e test.
Would be good to have the PR reviewed while I sort out the e2e test

**Comment by dchen1107** [comment]
@samabed Here are a list of e2e tests failed:

Test Result (11 failures / +8)
Kubernetes e2e suite.Kubectl client Update Demo should scale a replication controller
Kubernetes e2e suite.Services should be able to up and down services
Kubernetes e2e suite.Kubectl client Update Demo should create and stop a replication controller
Kubernetes e2e suite.Networking should function for intra-pod communication
Kubernetes e2e suite.Kubectl client Update Demo should do a rolling update of a replication controller
Kubernetes e2e suite.Proxy version v1 should proxy through a service and a pod
Kubernetes e2e suite.ReplicationController should serve a basic image on each replica with a public image
Kubernetes e2e suite.kube-ui should check that the kube-ui instance is alive
Kubernetes e2e suite.PreStop should call prestop when killing a pod
Kubernetes e2e suite.Services should be able to create a functioning NodePort service

**Comment by dchen1107** [comment]
ok to test

File: pkg/kubelet/dockertools/manager.go [hunk 1]
```diff
@@ -459,10 +459,17 @@ func (dm *DockerManager) GetPodStatus(pod *api.Pod) (*api.PodStatus, error) {
 		}
 	}
 
-	// Handle the containers for which we cannot find any associated active or
-	// dead docker containers.
+	// Handle the containers for which
```
Comment (dchen1107) [inline pkg/kubelet/dockertools/manager.go]: why move to next line here?

File: pkg/kubelet/dockertools/manager.go [hunk 2]
```diff
@@ -459,10 +459,17 @@ func (dm *DockerManager) GetPodStatus(pod *api.Pod) (*api.PodStatus, error) {
...
+	// we cannot find any associated active or dead docker containers.
+	// In backoff
```
Comment (dchen1107) [inline pkg/kubelet/dockertools/manager.go]: In complete sentence?

File: pkg/kubelet/dockertools/manager.go [hunk 3]
```diff
@@ -459,10 +459,17 @@ func (dm *DockerManager) GetPodStatus(pod *api.Pod) (*api.PodStatus, error) {
...
 	for _, container := range manifest.Containers {
-		if _, found := statuses[container.Name]; found {
+		if containerStatus, found := statuses[container.Name]; found {
+			reason, ok := dm.reasonCache.Get(uid, container.Name)
+			if ok && reason == kubecontainer.ErrCrashLoopBackOff.Error() {
+				containerStatus.LastTerminationState = containerStatus.State
+				containerStatus.State.Waiting = &api.ContainerStateWaiting{Reason: reason}
```
Comment (dchen1107) [inline pkg/kubelet/dockertools/manager.go]: Shouldn't reason for transferring ContainerState to Waiting is "Backoff" now?

**Comment by dchen1107** [comment]
LGTM overall except several nits.

File: pkg/kubelet/dockertools/manager.go (hunk 3 above)
Comment (samsabed) [inline pkg/kubelet/dockertools/manager.go]: reason (the variable) at this point contains ErrCrashLoopBackOff.Error(), which is fetched from the reasonCache.

**Comment by dchen1107** [comment]
LGTM now. Thanks for fixing this.

**Comment by zmerlynn** [comment]
Needs rebase now.

**Comment by samsabed** [comment]
rebased, I don't have access to see  the jenkins error.

**Comment by nikhiljindal** [comment]
@k8s-bot ok to test
It failed to fetch the repo last time.

**Pull request merged**